### PR TITLE
Switch HalfTraversal to APIv2

### PR DESCRIPTION
**Pull request by aprokop**
The main thing is to change half-traversal callback from `callback(int, int)` to `callback(value, value)`.

Note that this change is not backwards compatible. However, `HalfTraversal` is experimental, so it should be fine.

### CODE REVIEW

**Comment by aprokop** [comment]
I need to check that FDBSCAN is not affected. We don't have benchmark for neighbor lists, so can't say how much it affected it. But I expect that there is some benefit as I switched to APIv2 with points instead of boxes for the tree construction.

File: src/details/ArborX_NeighborList.hpp [hunk 1]
```diff
@@ -88,10 +102,21 @@ void findFullNeighborList(ExecutionSpace const &space,
...
+      KokkosExt::is_accessible_from<MemorySpace, ExecutionSpace>::value,
+      "Primitives must be accessible from the execution space");
+
+  using Point = typename Points::value_type;
+  static_assert(GeometryTraits::is_point<Point>{});
+
+  Points points{primitives};
+  int const n = points.size();
+
+  BoundingVolumeHierarchy<MemorySpace, PairValueIndex<Point>> bvh(
```
Comment (aprokop) [inline src/details/ArborX_NeighborList.hpp:118]: This should improve performance.

File: src/details/ArborX_NeighborList.hpp [hunk 2]
```diff
@@ -55,7 +67,9 @@ void findHalfNeighborList(ExecutionSpace const &space,
   Kokkos::deep_copy(space, offsets, 0);
   HalfTraversal(
       space, bvh,
-      KOKKOS_LAMBDA(int, int j) { Kokkos::atomic_increment(&offsets(j)); },
+      KOKKOS_LAMBDA(auto, auto const &value) {
```
Comment (dalg24) [inline src/details/ArborX_NeighborList.hpp:70]: Generic lambdas are not supported by NVCC

File: src/ArborX_DBSCAN.hpp [hunk 3]
```diff
@@ -63,14 +64,17 @@ struct WithinRadiusGetter
 {
   float _r;
 
-  template <typename Point>
-  KOKKOS_FUNCTION auto operator()(Point const &point) const
+  template <typename Geometry, typename Index>
+  KOKKOS_FUNCTION auto
+  operator()(PairValueIndex<Geometry, Index> const &value) const
```
Comment (dalg24) [inline src/ArborX_DBSCAN.hpp:69]: Don't call the pair (value,index) "value".  This is confusing.

**Comment by aprokop** [comment]
Move a part of the PR (the one that isn't related to switching HalfTraversal API to v2) to #1024.

**Comment by aprokop** [comment]
Rebased on #1024.

**Comment by aprokop** [comment]
The following change seems to have fixed the performance issue:
```diff
--- a/src/details/ArborX_DetailsHalfTraversal.hpp
+++ b/src/details/ArborX_DetailsHalfTraversal.hpp
@@ -52,7 +52,7 @@ struct HalfTraversal

   KOKKOS_FUNCTION void operator()(int i) const
   {
-    auto const &leaf_value = HappyTreeFriends::getValue(_bvh, i);
+    auto const leaf_value = HappyTreeFriends::getValue(_bvh, i);
     auto const predicate = _get_predicate(leaf_value);

     int node = HappyTreeFriends::getRope(_bvh, i);
```
While it may have negative consequences for very large size values, this is acceptable for our common use cases. In the future, we may specialize on `sizeof(Value)`.

```
saturn dbscan ((fd39c111...)) $ for i in ./ArborX_master_466efb7c ./ArborX_half_fd39c111; do $i --binary --filename ~/data/hacc_497M.arborx --eps 0.014 --core-min-size 2 --impl fdbscan --verbose --kokkos-device-id=2 --max-num-points 300000000; done
ArborX version    : 1.6 (dev)
ArborX hash       : 466efb7c
Kokkos version    : 4.1.0
algorithm         : dbscan
eps               : 0.014000
cluster min size  : 1
implementation    : fdbscan
verify            : false
minpts            : 2
filename          : /home/users/aprokop/data/hacc_497M.arborx [binary, max_pts = 300000000]
samples           : -1
verbose           : true
Reading in "/home/users/aprokop/data/hacc_497M.arborx" in binary mode...done
Read in 300000000 3D points
-- construction     :      0.207
-- query+cluster    :      1.894
-- postprocess      :      0.202
total time          :      2.308

#clusters       : 18287721
#cluster points : 226764477 [75.59%]
#noise   points : 73235523 [24.41%]
ArborX version    : 1.6 (dev)
ArborX hash       : fd39c111
Kokkos version    : 4.1.0
algorithm         : dbscan
eps               : 0.014000
cluster min size  : 1
implementation    : fdbscan
verify            : false
minpts            : 2
filename          : /home/users/aprokop/data/hacc_497M.arborx [binary, max_pts = 300000000]
samples           : -1
verbose           : true
Reading in "/home/users/aprokop/data/hacc_497M.arborx" in binary mode...done
Read in 300000000 3D points
-- construction     :      0.242
-- query+cluster    :      1.882
-- postprocess      :      0.203
total time          :      2.332

#clusters       : 18287721
#cluster points : 226764477 [75.59%]
#noise   points : 73235523 [24.41%]
saturn dbscan ((fd39c111...)) $ for i in ./ArborX_master_466efb7c ./ArborX_half_fd39c111; do $i --binary --filename ~/data/hacc_497M.arborx --eps 0.014 --core-min-size 10 --impl fdbscan --verbose --kokkos-device-id=2 --max-num-points 300000000; done
ArborX version    : 1.6 (dev)
ArborX hash       : 466efb7c
Kokkos version    : 4.1.0
algorithm         : dbscan
eps               : 0.014000
cluster min size  : 1
implementation    : fdbscan
verify            : false
minpts            : 10
filename          : /home/users/aprokop/data/hacc_497M.arborx [binary, max_pts = 300000000]
samples           : -1
verbose           : true
Reading in "/home/users/aprokop/data/hacc_497M.arborx" in binary mode...done
Read in 300000000 3D points
-- construction     :      0.241
-- query+cluster    :      2.641
---- neigh          :      0.449
---- query          :      2.098
-- postprocess      :      0.146
total time          :      3.034

#clusters       : 791387
#cluster points : 148503760 [49.50%]
#noise   points : 151496240 [50.50%]
ArborX version    : 1.6 (dev)
ArborX hash       : fd39c111
Kokkos version    : 4.1.0
algorithm         : dbscan
eps               : 0.014000
cluster min size  : 1
implementation    : fdbscan
verify            : false
minpts            : 10
filename          : /home/users/aprokop/data/hacc_497M.arborx [binary, max_pts = 300000000]
samples           : -1
verbose           : true
Reading in "/home/users/aprokop/data/hacc_497M.arborx" in binary mode...done
Read in 300000000 3D points
-- construction     :      0.179
-- query+cluster    :      2.643
---- neigh          :      0.448
---- query          :      2.104
-- postprocess      :      0.146
total time          :      2.973

#clusters       : 791387
#cluster points : 148503760 [49.50%]
#noise   points : 151496240 [50.50%]
```

File: src/details/ArborX_NeighborList.hpp (hunk 2 above)
Comment (dalg24) [reply src/details/ArborX_NeighborList.hpp:80]: Same comment still applies.

**Comment by aprokop** [comment]
@dalg24 ready for re-review

File: src/details/ArborX_NeighborList.hpp [hunk 4]
```diff
@@ -122,7 +127,9 @@ void findFullNeighborList(ExecutionSpace const &space,
   Points points{primitives}; // NOLINT
   int const n = points.size();
 
-  BoundingVolumeHierarchy<MemorySpace, PairValueIndex<Point>> bvh(
+  using Value = PairValueIndex<Point>;
```
Comment (dalg24) [inline src/details/ArborX_NeighborList.hpp:130]: Not a big fan of that naming

**Review by dalg24 — APPROVED** [review]
I am a bit surprised by the variability of the construction timings in the results you posted.  Is it something you have been seeing before?

**Comment by aprokop** [comment]
> I am a bit surprised by the variability of the construction timings in the results you posted. Is it something you have been seeing before?

Something else was running which may have interfered at the moment. I rerun today a few times and saw consistent 0.207-0.210.

**Pull request merged**